File: Linguagem de expressão 1/src/main/java/le1/plp/expressions1/expression/ExpAnd.java
package le1.plp.expressions1.expression;

import le1.plp.expressions1.util.Tipo;

/**
 * Um objeto desta classe representa uma Expressao de Conjuncao logica.
 */
public class ExpAnd extends ExpBinaria{

	/**
	 * Controi uma Expressao de Conjuncao logica  com as sub-expressoes
	 * especificadas. Estas devem ser tais que sua avaliacao retorna
	 * <code>ValorBooleano</code>
	 *
	 * @param esq Expressao da esquerda
	 * @param dir Expressao da direita
	 */
	public ExpAnd(Expressao esq, Expressao dir) {
		super(esq, dir, "and");
	}


	/**
	 * Retorna o valor da Expressao de Conjuncao Logica
	 */
	 public Valor avaliar(){
		return new ValorBooleano(
					((ValorBooleano)getEsq().avaliar()).valor() &&
					((ValorBooleano)getDir().avaliar()).valor()
		);
	}
 
	/**
	 * Realiza a verificacao de tipos desta expressao.
	 *
	 * @return <code>true</code> se os tipos da expressao sao validos;
	 *          <code>false</code> caso contrario.
	 */
	protected boolean checaTipoElementoTerminal() {
		return (getEsq().getTipo().eBooleano() && getDir().getTipo().eBooleano());
	}

	/**
	 * Retorna os tipos possiveis desta expressao.
	 *
	 * @return os tipos possiveis desta expressao.
	 */
	public Tipo getTipo() {
		return Tipo.TIPO_BOOLEANO;
	}

}

File: Le1 - Python/le1/expression/exp_and.py
from __future__ import annotations

from le1.expression.exp_binaria import ExpBinaria
from le1.expression.expressao import Expressao
from le1.expression.valor_booleano import ValorBooleano
from le1.util.tipo import Tipo


class ExpAnd(ExpBinaria):
    def __init__(self, esq: Expressao, dir: Expressao) -> None:
        super().__init__(esq, dir, "and")

    def avaliar(self) -> ValorBooleano:
        valor_esq = self.getEsq().avaliar()
        valor_dir = self.getDir().avaliar()

        if not isinstance(valor_esq, ValorBooleano):
            raise TypeError("ExpAnd espera ValorBooleano à esquerda")

        if not isinstance(valor_dir, ValorBooleano):
            raise TypeError("ExpAnd espera ValorBooleano à direita")

        return ValorBooleano(valor_esq.valor() and valor_dir.valor())

    def checaTipoElementoTerminal(self) -> bool:
        return self.getEsq().getTipo().eBooleano() and self.getDir().getTipo().eBooleano()

    def getTipo(self) -> Tipo:
        return Tipo.TIPO_BOOLEANO